### PR TITLE
port lane_departure_checker

### PR DESCRIPTION
**Pull request by rej55**
## Description

Porting lane_departure_checker for using autoware_auto_msgs.
- Use autoware_auto_planning_msgs, autoware_auto_mapping_msgs.
- Use nav_msgs/Odometry to get current twist and pose covariance.
- Use TrajectoryPointArray for resampling

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [pull request guidelines][pull-request-guidelines]
- [x] Code follows [coding guidelines][coding-guidelines]
- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code follows [coding guidelines][coding-guidelines]
- [ ] (Optional) Unit tests have been written for new behavior
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [x] All open points are addressed and tracked via issues or tickets
- [x] Write [release notes][release-notes]

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[clang-tidy-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/ClangTidyGuideline/
[coding-guidelines]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/1194394777/T4
[pull-request-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/PullRequestGuideline/
[release-notes]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/563774416
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp [hunk 1]
```diff
@@ -37,8 +39,12 @@
 
 namespace lane_departure_checker
 {
+using autoware_auto_planning_msgs::msg::HADMapRoute;
+using autoware_auto_planning_msgs::msg::Trajectory;
+using autoware_auto_planning_msgs::msg::TrajectoryPoint;
 using autoware_utils::LinearRing2d;
 using autoware_utils::PoseDeviation;
+using TrajectoryPointArray = std::vector<autoware_auto_planning_msgs::msg::TrajectoryPoint>;
```
Comment (kenji-miyake) [inline control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp:47]: I feel keeping `std::vector` (`std::vector<TrajectoryPoint>`) is more readable. How about removing `using TrajectoryPointArray`?

File: control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker_node.hpp [hunk 2]
```diff
@@ -39,6 +41,12 @@
 
 namespace lane_departure_checker
 {
+using autoware_auto_mapping_msgs::msg::HADMapBin;
+using autoware_auto_planning_msgs::msg::HADMapRoute;
+using autoware_auto_planning_msgs::msg::Trajectory;
+using autoware_auto_planning_msgs::msg::TrajectoryPoint;
+using TrajectoryPointArray = std::vector<autoware_auto_planning_msgs::msg::TrajectoryPoint>;
```
Comment (kenji-miyake) [inline control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker_node.hpp:48]: Why did you write this here and in `lane_departure_checker.hpp`?
I guess `lane_departure_checker_node.hpp` includes `lane_departure_checker.hpp` and no need to write here again.

File: control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp (hunk 1 above)
Comment (rej55) [reply control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp:47]: When we port planning packages, planning & control team agreed that we use this.
@TakaHoribe How do you think?

File: control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker_node.hpp [hunk 3]
```diff
@@ -53,35 +61,31 @@ class LaneDepartureCheckerNode : public rclcpp::Node
...
+  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_odom_;
+  rclcpp::Subscription<HADMapBin>::SharedPtr sub_lanelet_map_bin_;
+  rclcpp::Subscription<HADMapRoute>::SharedPtr sub_route_;
+  rclcpp::Subscription<Trajectory>::SharedPtr sub_reference_trajectory_;
+  rclcpp::Subscription<Trajectory>::SharedPtr sub_predicted_trajectory_;
 
   // Data Buffer
+  nav_msgs::msg::Odometry::ConstSharedPtr current_odom_;
   geometry_msgs::msg::PoseStamped::ConstSharedPtr current_pose_;
-  geometry_msgs::msg::TwistStamped::ConstSharedPtr current_twist_;
```
Comment (kenji-miyake) [inline control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker_node.hpp:67]: [nits] Why did you change the order (to before `current_pose_`)? The diff looks a bit strange.

File: control/trajectory_follower/lane_departure_checker/launch/lane_departure_checker.launch.xml [hunk 4]
```diff
@@ -1,19 +1,17 @@
...
   <arg name="config_file" default="$(find-pkg-share lane_departure_checker)/config/lane_departure_checker.param.yaml" />
 
   <node pkg="lane_departure_checker" exec="lane_departure_checker_node" name="lane_departure_checker_node" output="screen">
-    <remap from="~/input/twist" to="$(var input/twist)"/>
+    <remap from="~/input/odometry" to="$(var input/odometry)"/>
     <remap from="~/input/lanelet_map_bin" to="$(var input/lanelet_map_bin)"/>
     <remap from="~/input/route" to="$(var input/route)"/>
     <remap from="~/input/reference_trajectory" to="$(var input/reference_trajectory)"/>
     <remap from="~/input/predicted_trajectory" to="$(var input/predicted_trajectory)"/>
-    <remap from="~/input/covariance" to="$(var input/covariance)"/>
```
Comment (kenji-miyake) [inline control/trajectory_follower/lane_departure_checker/launch/lane_departure_checker.launch.xml:16]: @k0suke-murakami Could you check around `covariance` logic, please? :pray:

File: control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp (hunk 1 above)
Comment (kenji-miyake) [reply control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp:47]: Since there is `geometry_msgs/msg/PoseArray`, it looks like a msg to me.
IMO, `TrajectoryPoints` is better and acceptable.

Comment (TakaHoribe) [reply control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp:47]: `std::vector<TrajectoryPoint>` は今後型が変わる可能性があると思ったのと、コードが vectorまみれになるのを懸念してusing勧めました。

usingにデメリットがあるのであれば 消してしまっても良いです！
か、三宅さん的には `TrajectoryPoints` でも良い感じでしょうか？

Comment (TakaHoribe) [reply control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp:47]: （英語必須でしたっけ・・・？）

Comment (kenji-miyake) [reply control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp:47]: Maximeさんも一緒にやっているのと、今後AWFのメンバーが議論の過程を見るかもしれないので、英語にした方が良いのかなぁ、とは思っています。（あとは我々の練習として…）

> usingにデメリットがあるのであれば 消してしまっても良いです！
か、三宅さん的には TrajectoryPoints でも良い感じでしょうか？

はい、`TrajectoryPoints` の方がvector感があって僕は好きですね。あと短いですし。
まぁ慣れの問題だとは思うので、他パッケージとの整合性がある程度取れていれば何でも大丈夫な感じはします。

Comment (TakaHoribe) [reply control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp:47]: Cool. I love that idea.
Ok, then let's use the trajectoryPoints.

File: control/trajectory_follower/lane_departure_checker/launch/lane_departure_checker.launch.xml (hunk 4 above)
Comment (k0suke-murakami) [reply control/trajectory_follower/lane_departure_checker/launch/lane_departure_checker.launch.xml:16]: Looks fine to me.  (From what I see from diff, there is no change in core covariance logic.)

File: control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp (hunk 1 above)
Comment (rej55) [reply control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker.hpp]: I changed the alias.
https://github.com/tier4/AutowareArchitectureProposal.iv/pull/499/commits/4c7a50050fa8e70b19f2b11a8d486ae807df344b

File: control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker_node.hpp (hunk 2 above)
Comment (rej55) [reply control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker_node.hpp]: I removed unnecessary alias
https://github.com/tier4/AutowareArchitectureProposal.iv/pull/499/commits/4c7a50050fa8e70b19f2b11a8d486ae807df344b

File: control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker_node.hpp (hunk 3 above)
Comment (rej55) [reply control/trajectory_follower/lane_departure_checker/include/lane_departure_checker/lane_departure_checker_node.hpp:67]: There is no reason to change the order. I fixed.
https://github.com/tier4/AutowareArchitectureProposal.iv/pull/499/commits/4c7a50050fa8e70b19f2b11a8d486ae807df344b

**Review by kenji-miyake — APPROVED** [review]
